### PR TITLE
Improve namespace filtering

### LINKED ISSUE #15: Improve namespace allow/block system
## Brief

Improve the way to make the operator monitor namespaces with an easier, more "semantic" and less error-prone way.  
Note: this is not an urgent issue, but can improve usability of the operator. Code-wise it won't be hard to change, but depending on how users view the current system (easy or confusing) this maybe long or short term.

## Current way of doing it

An administrator/user can define which namespaces the operator can monitor by putting namespaces inside an *allowlist* or *blocklist*.

If settings are set as `listPolicy: allowlist` then the operator will only monitor namespaces that have label `operator.cnwan.io/allowed: yes` (but really the value can be anything, more on this later).  
On the contrary, `listPolicy: blocklist` will make the operator monitor all namespaces and avoid those that have label `operator.cnwan.io/blocked: yes` (again, the value can really be anything here).

## Issue

This system is very convenient but can certainly be improved, as it has some shady parts that can be considered bad practice or misleading.

### Labels

* We have *two* labels to filter a namespace (maybe filter is not the right word here, but it explains weel), instead of one. What happens when a user - by mistake - puts both labels `operator.cnwan.io/allowed` and `operator.cnwan.io/blocked` in the same namespace? The operator will know which one is the one that it has two check. But nonetheless this can confuse users, who can potentially mis-configure the operator, which in turn can have unwanted/unexpected behaviors.
  * Users may be tempted to do `operator.cnwan.io/allowed: no`. But this will have no effect as the key is actually the important part (read next point) and will bring even more confusion.

* `operator.cnwan.io/allowed: yes` and `operator.cnwan.io/blocked: yes` are a bad use of labels and a bad practice as a consequence: in labels and metadata what is actually meaningful is the *value*, **not** the key. They key is supposed to be the differentiator and *holder* (after all, you can think of it as a YAML key), while the value is supposed be the important thing because it tells how key must be configured. Instead, we are actually considering the key here (*../allowed* or *../blocked*) as the **actual** value, which is a bad practice because it is not what its job is.
  * As a consequence, the value is really meaningless because it can be whatever thing since we consider the key to be important. Making it more meaningful by forcing it to have `yes` or `true` is useless, because in that case it might just as well not be there entirely. Additionally, I think it would be pointless to have a key that can have only two values (synonyms) and the others are "invalid" (for lack of other word) in light of the previous sentence.

* Minor thing as it is just a language issue: but the operator is not really allowed or blocked to do anything here so it is just a minor confusion point.

### List Policy

The current way to set the default list as `listPolicy: allowlist/blocklist` is good and can even stay like this, but I think that it can be improved and made easier to understand.

* As said above, there is no real allow/block here as there is no real concept of "permissions" that you can give to the operator.
* The way it works can be re-worked in a easier way.
* Minor thing: is "policy" a good/fitting word here?
* Minor thing (but it may just be me): when re-writing the documentation and describing this part, I had the feeling that `listPolicy: allowlist` may be wrongly interpreted as "namespaces are in the allowlist by default" if one does not read the documentation.

## Changes

I intend/propose to change the operator to make it better and more understandable with following ways.

### From ListPolicy to DefaultBehavior

The `listPolicy` setting can be moved to a better description.

For example: `listPolicy` can be changed to `defaultBehavior` or `defaultAction` to make it clear that it modifies the default way the operator works.

* `listPolicy: allowlist` can be changed to `defaultBehavior: ignore` which makes the operator **ignore** (I think this is a better fitting word) a namespace *by default*.
* `listPolicy: blocklist` can be changed to `defaultBehavior: watch` which makes the operator **watch** (again, better fitting word) a namespace *by default*.

### Labels

This is more pressing issue.

* Instead of having two different keys, have only one that is more good practice: `operator.cnwan.io/action`. The prefix `operator.cnwan.io` again proves that what comes next (`action`) only makes sense for the cnwan operator (just as `k8s.io/something` makes sense for kubernetes internal purposes).
  * There is no risk of having two keys and maybe messing up something.

* To watch a namespace use values `watch`, `include`, `enable` and maybe something else. I think that `operator.cnwan.io/action: watch` is more fitting and better understandable. `watch` is of course enough, but multiple synonyms allow a user not to have to look at the documentation if they don't remember the valid word.
* On the contrary, `ignore`, `disable`, `exclude`, etc. will replace `operator.cnwan.io/blocked`.

## Upgrade

For people that used versions where the old names were used, an `upgrade.sh` shell be provided that will automatically write the correct values and change the labels accordingly.
A *DEPRECATION NOTICE* will be included in the documentation.

## Support

The old mechanics will be supported for x (like, 2?) major version cycles with the `DEPRECATED` notice.

### PR DESCRIPTION
**Pull request by SunSince90**
## TL;DR

- New setting `enableNamespaceByDefault`: 
  - if false => ignore namespace
  - if true => watch namespace
- New label `operator.cnwan.io/enabled=yes|no`.
  - `yes`: watch the namespace
  - `no`: stay away from namespace
  - name is very fitting and can be used to filter other resources in future (e.g. enable *this* namespace but disable *this* service under it)

## Description

This PR closes #15 and improves the current way a namespace is blocked/allowed. Read the issue for more information, although the changes have been applied in a slightly different way.

This provides a much simpler way to filter namespaces that must be ignored or watched by the operator: there is no more concept of *allowlist* or *blocklist*, and there is a more descriptive label.

You just need to label the namespace as `operator.cnwan.io/enabled` with either `yes` and the operator will watch service updates on this namespace or `no` and the operator will stay away from it. This replaces the old `operator.cnwan.io/allowed` and `operator.cnwan.io/blocked` as they were redundant and made the value irrelevant (it was basically ignored). Again read the issue for more information about why I consider it a bad practice.

If such a label is not found, then the operator will pretend it saw `operator.cnwan.io/enabled=no`, unless `enableNamespaceByDefault` is set to true on the operator's settings, in which case it will pretend it saw `operator.cnwan.io/enabled=yes`.

Finally, this fixes some other minor bugs found while developing the feature.

## Status

- [x] Replace old `namespaceListPolicy` setting with a simple `enableNamespaceByDefault` boolean one
- [x] Replace old labels with `operator.cnwan.io/enabled` labels
- [x] Update controllers
- [x] Update unit tests
- [x] Update documentation

### CODE REVIEW

**Comment by SunSince90** [comment]
As usual, PR will exit draft once all checks are marked as completed

**Comment by SunSince90** [comment]
I just renamed all instances of "enable" with "monitor", as we previously discussed. Although now I am (over-)thinking if maybe "watch" would be better, to be more consistent :D :D

**Comment by ljakab** [comment]
> I just renamed all instances of "enable" with "monitor", as we previously discussed. Although now I am (over-)thinking if maybe "watch" would be better, to be more consistent :D :D

Indeed, "watch" sounds better to me too, let's see what @arnatal has to say... :)

File: docs/concepts.md [hunk 1]
```diff
@@ -81,47 +80,31 @@ That being said, the operator will still insert child resources even if the pare
...
-# Insert namespace in the blocklist
-kubectl label ns namespace-name operator.cnwan.io/blocked=yes
-```
+Instead, if you have many namespaces to monitor and/or find it tedious to manually do so for every single one of them, you can override this behavior via `monitorNamespacesByDefault: true` on the [operator settings](./configuration.md#monitor-namespaces-by-default): this means that the operator will pretend it is seeing `operator.cnwan.io/monitor=true` and thus watch events in the namespace by default, unless instructed otherwise. This is the opposite scenario from above: now you will need to manually *disable* them.
 
-It doesn't really matter what you put as value (in this case `yes` has been inserted), just as long as the key `operator.cnwan.io/<key>` is as specified above.
+Let's see some examples.
 
-Similarly, to remove a namespace from the list:
+To _monitor_ monitoring on namespace `hr`, do the following:
```
Comment (ljakab) [inline docs/concepts.md:95]: I suppose this is a reasult of search and replace "To _enable_ monitoring". I guess you can leave "To _monitor_" or change back "To _enable_ monitoring" to be consistent with the next example below about disabling monitoring.

**Review by ljakab — CHANGES_REQUESTED** [review]

File: docs/concepts.md (hunk 1 above)
Comment (SunSince90) [reply docs/concepts.md:95]: Exactly, thank you for catching this!

Comment (SunSince90) [reply docs/concepts.md]: Solved by amending to

> _enable_ monitoring

**Review by ljakab — APPROVED** [review]
